Add lazyZip operation (formerly zipWith)

marcelocenerine opened the pull request:
Changes made for #221:
- Add `zipWith` method to `collection.IterableOps` and `collection.Iterator`
- Introduce view `ZipWith`
- Override `zipWith` method in `ImmutableArray` and `LazyList` to be consistent with the `zip` operation
- Overload `zipWith` method in `collection.ArrayOps` and `collection.SortedSet` to be consistent with the `zip` operation
- Introduce benchmarks for `zipWith`, `zip` and `zipWithIndex` (the latter two were missing)

It was not clear to me how the unit tests are organized. I wrote a bunch of them for the main collection types in both immutable and mutable packages to make sure my implementation works  but did not include them in this PR.

julienrf reviewed (commented):
What’s the impact on performance if we implement `zip` in terms of `zipWith`, to avoid some code duplication?

marcelocenerine commented:
I was doing that at this very moment :D. It's gonna take some time until the benchmarks finish. I'll keep you posted

EPronovost commented:
Oh well, I implemented this as well and literally wrote the exact same code. Nice work! 
 https://github.com/EPronovost/collection-strawman/commit/24049b57bcc8d18ad1f3b554d8f39f0c997f7181

marcelocenerine commented:
Note:
- `zip` == existing implementation
- `zip2` == `zip` implemented in terms of `zipWith`

![zip_vs_zip2](https://user-images.githubusercontent.com/1107367/29997647-3f0591b4-9010-11e7-9a44-f7ceea351462.png)

```
Benchmark               (size)  Mode  Cnt      Score      Error  Units
HashSetBenchmark.zip         0  avgt   12     19.750 ±    1.138  ns/op
HashSetBenchmark.zip         1  avgt   12     65.216 ±    3.590  ns/op
HashSetBenchmark.zip         2  avgt   12    150.916 ±    8.569  ns/op
HashSetBenchmark.zip         3  avgt   12    223.390 ±    3.237  ns/op
HashSetBenchmark.zip         4  avgt   12    321.306 ±    3.789  ns/op
HashSetBenchmark.zip         7  avgt   12    592.327 ±   17.823  ns/op
HashSetBenchmark.zip         8  avgt   12    682.168 ±   35.707  ns/op
HashSetBenchmark.zip        15  avgt   12   1280.184 ±   70.554  ns/op
HashSetBenchmark.zip        16  avgt   12   1393.507 ±   48.281  ns/op
HashSetBenchmark.zip        17  avgt   12   1496.845 ±   68.302  ns/op
HashSetBenchmark.zip        39  avgt   12   4762.469 ±  295.681  ns/op
HashSetBenchmark.zip       282  avgt   12  44191.198 ± 2168.640  ns/op

HashSetBenchmark.zip2        0  avgt   12     20.369 ±    0.590  ns/op
HashSetBenchmark.zip2        1  avgt   12     60.988 ±    1.465  ns/op
HashSetBenchmark.zip2        2  avgt   12    147.952 ±    5.945  ns/op
HashSetBenchmark.zip2        3  avgt   12    236.831 ±    4.997  ns/op
HashSetBenchmark.zip2        4  avgt   12    335.372 ±   14.477  ns/op
HashSetBenchmark.zip2        7  avgt   12    551.376 ±    7.048  ns/op
HashSetBenchmark.zip2        8  avgt   12    651.141 ±   20.254  ns/op
HashSetBenchmark.zip2       15  avgt   12   1251.230 ±   36.530  ns/op
HashSetBenchmark.zip2       16  avgt   12   1436.768 ±   38.667  ns/op
HashSetBenchmark.zip2       17  avgt   12   1532.003 ±   20.423  ns/op
HashSetBenchmark.zip2       39  avgt   12   4728.074 ±  174.185  ns/op
HashSetBenchmark.zip2      282  avgt   12  44729.714 ± 1118.149  ns/op
```

```
Benchmark               (size)  Mode  Cnt      Score      Error  Units
ListBenchmark.zip            0  avgt   12     92.196 ±    3.679  ns/op
ListBenchmark.zip            1  avgt   12    103.485 ±    1.883  ns/op
ListBenchmark.zip            2  avgt   12    162.533 ±   12.149  ns/op
ListBenchmark.zip            3  avgt   12    193.258 ±   14.573  ns/op
ListBenchmark.zip            4  avgt   12    208.044 ±    8.488  ns/op
ListBenchmark.zip            7  avgt   12    304.174 ±    8.090  ns/op
ListBenchmark.zip            8  avgt   12    276.759 ±    6.379  ns/op
ListBenchmark.zip           15  avgt   12    536.954 ±   14.958  ns/op
ListBenchmark.zip           16  avgt   12    421.065 ±   12.324  ns/op
ListBenchmark.zip           17  avgt   12    383.749 ±    7.729  ns/op
ListBenchmark.zip           39  avgt   12    809.668 ±   18.085  ns/op
ListBenchmark.zip          282  avgt   12   5699.023 ±  100.800  ns/op

ListBenchmark.zip2           0  avgt   12     90.923 ±    1.341  ns/op
ListBenchmark.zip2           1  avgt   12    103.967 ±    1.503  ns/op
ListBenchmark.zip2           2  avgt   12    130.143 ±    3.979  ns/op
ListBenchmark.zip2           3  avgt   12    153.424 ±    7.152  ns/op
ListBenchmark.zip2           4  avgt   12    178.389 ±    2.642  ns/op
ListBenchmark.zip2           7  avgt   12    246.142 ±    4.817  ns/op
ListBenchmark.zip2           8  avgt   12    276.040 ±    7.982  ns/op
ListBenchmark.zip2          15  avgt   12    429.587 ±   17.765  ns/op
ListBenchmark.zip2          16  avgt   12    373.032 ±    4.830  ns/op
ListBenchmark.zip2          17  avgt   12    388.918 ±    6.691  ns/op
ListBenchmark.zip2          39  avgt   12    983.104 ±   31.814  ns/op
ListBenchmark.zip2         282  avgt   12   5705.923 ±  104.434  ns/op
```

```
Benchmark               (size)  Mode  Cnt      Score      Error  Units
LazyListBenchmark.zip        0  avgt   12      3.744 ±    0.084  ns/op
LazyListBenchmark.zip        1  avgt   12     14.669 ±    0.444  ns/op
LazyListBenchmark.zip        2  avgt   12     15.487 ±    0.673  ns/op
LazyListBenchmark.zip        3  avgt   12     19.166 ±    3.590  ns/op
LazyListBenchmark.zip        4  avgt   12     17.731 ±    1.842  ns/op
LazyListBenchmark.zip        7  avgt   12     16.486 ±    2.188  ns/op
LazyListBenchmark.zip        8  avgt   12     16.813 ±    2.053  ns/op
LazyListBenchmark.zip       15  avgt   12     15.074 ±    0.599  ns/op
LazyListBenchmark.zip       16  avgt   12     15.743 ±    0.579  ns/op
LazyListBenchmark.zip       17  avgt   12     15.599 ±    0.696  ns/op
LazyListBenchmark.zip       39  avgt   12     17.200 ±    2.322  ns/op
LazyListBenchmark.zip      282  avgt   12     16.031 ±    1.493  ns/op

LazyListBenchmark.zip2       0  avgt   12      4.113 ±    0.097  ns/op
LazyListBenchmark.zip2       1  avgt   12     17.683 ±    1.195  ns/op
LazyListBenchmark.zip2       2  avgt   12     16.369 ±    0.315  ns/op
LazyListBenchmark.zip2       3  avgt   12     16.265 ±    0.257  ns/op
LazyListBenchmark.zip2       4  avgt   12     16.310 ±    0.445  ns/op
LazyListBenchmark.zip2       7  avgt   12     17.288 ±    0.496  ns/op
LazyListBenchmark.zip2       8  avgt   12     17.595 ±    0.359  ns/op
LazyListBenchmark.zip2      15  avgt   12     17.463 ±    0.899  ns/op
LazyListBenchmark.zip2      16  avgt   12     17.239 ±    0.703  ns/op
LazyListBenchmark.zip2      17  avgt   12     17.149 ±    0.268  ns/op
LazyListBenchmark.zip2      39  avgt   12     17.236 ±    0.349  ns/op
LazyListBenchmark.zip2     282  avgt   12     17.160 ±    0.734  ns/op
```

@julienrf, should I go ahead and update the PR or keep them as is?

sjrd commented:
IMO `zipWith` is a terrible name for this operation. Naming-wise, it's basically equivalent to `zip`. Two reasonable names, IMO, would be `mapWith` (because you are mapping this collection together with another one) or `zipMap` (because, obviously, it's a `zip` followed by a `map`).

marcelocenerine commented:
`zipWith` would be [consistent with Haskell](http://hackage.haskell.org/package/base-4.10.0.0/docs/Data-List.html#v:zipWith) though

sjrd commented:
In Haskell it makes more sense because the mapping function comes first (due to currying conventions), which means that `zipWith (+) xs ys` reads as "zip with operation `+` the lists `xs` and `ys`. It also makes sense because you can define `let zip = zipWith (+)` (basically; I'm not sure of my Haskell syntax).

In Scala, however, `xs.zipWith(ys)(_ + _)` reads as "zip `xs` with `ys`, and some operation". But that would also be the reading of `xs.zip(ys)`.

Precedents in Scala that would match Haskell's way of thinking use `By`: `sortBy`, `groupBy`, etc. So I guess `zipBy` would fit the consistency with other Scala criterium, but I'm not sure it makes much sense as a reading on its own.

Ichoran commented:
I think we need to call it `zipTo`.  The problem with `zipWith` is that it evokes `zipWithIndex` which suggests that `zipWithIndex` is a specialized case of `zipWith`, which it's not.

smarter commented:
> The problem with zipWith is that it evokes zipWithIndex which suggests that zipWithIndex is a specialized case of zipWith, which it's not.

What do you mean by that? In Haskell zipWithIndex can be implemented as a special case of zip (which is a special case of zipWith):
```haskell
zipWithIndex xs = zip xs [0..((length xs) - 1)]
```

sjrd commented:
The `With` in `zipWithIndex` does not mean the same thing as the `With` in `zipWith`. The former introduces the rhs collection; the latter introduces the mapping function. At the end of the day, it's similar to my earlier argument: the reading of `zipWith` in Scala is such that you expect it to introduce the rhs collection, not the mapping function.

ritschwumm commented:
if you call this ```zipBy```, how would you call a function like this? or does that even exist already under a different name?
```scala
trait Coll[A] {
  def zipBy[B](f:A=>B):Coll[(A,B)] = this map { it => (it, f(it)) }
}
```

julienrf commented:
Thanks @marcelocenerine for running the benchmarks :) For me it makes sense to implement `zip` in terms of `zipWith`.

About the name… So far the proposals have been the following:

~~~ scala
xs.zipTo(ys)(_ + _)
xs.zipBy(ys)(_ + _)
xs.zipMap(ys)(_ + _)
xs.zipWith(ys)(_ + _)
xs.mapWith(ys)(_ + _)
~~~

It’s unfortunate that the “with” doesn’t introduces the same thing as in Haskell. Therefore I don’t see a strong reason to follow the same naming convention. However the name `zipWith` is already established in other Scala libraries (e.g. cats, scalaz, akka-stream). Last but not least, in the Scala standard library already has `zipWith` for `Future` ([here](http://scala-lang.org/api/current/scala/concurrent/Future.html#zipWith[U,R](that:scala.concurrent.Future[U])(f:(T,U)=%3ER)(implicitexecutor:scala.concurrent.ExecutionContext):scala.concurrent.Future[R])).

marcelocenerine commented:
Sounds good, @julienrf. I cherry-picked the commit and updated this PR

julienrf commented on src/main/scala/strawman/collection/immutable/LazyList.scala:
```diff
@@ -172,12 +172,13 @@ sealed abstract class LazyList[+A]
     }
 
-  override final def zip[B](xs: collection.Iterable[B]): LazyList[(A, B)] =
+  override final def zip[B](xs: collection.Iterable[B]): LazyList[(A, B)] = zipWith(xs)((_, _))
```
Why do we need to override it here?

marcelocenerine replied:
good catch! After rewriting `zip` in terms of `zipWith`, there's no need to override `zip` here.

julienrf commented on src/main/scala/strawman/collection/immutable/ImmutableArray.scala:
```diff
@@ -76,14 +76,16 @@ class ImmutableArray[+A] private[collection] (private val elements: scala.Array[
     }
 
-  override def zip[B](xs: collection.Iterable[B]): ImmutableArray[(A, B)] =
+  override def zip[B](xs: collection.Iterable[B]): ImmutableArray[(A, B)] = zipWith(xs)((_, _))
```
Why is this override necessary?

marcelocenerine replied:
same here

julienrf commented:
Thanks @marcelocenerine. Could you add some tests too? (in the `test/junit` project)

marcelocenerine commented:
@julienrf, my apologies for taking this long to reply to your review comments (I haven't had access to my laptop for a few days). I just pushed a new commit addressing them.

Regarding the method name: ReactiveX (see [rxscala](http://reactivex.io/rxscala/scaladoc/index.html#rx.lang.scala.Observable)) is another library where `zipWith` is already established.

julienrf commented on src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -835,9 +835,24 @@ trait IterableOps[+A, +CC[X], +C] extends Any {
-  def zip[B](xs: Iterable[B]): CC[(A @uncheckedVariance, B)] = fromIterable(View.Zip(toIterable, xs))
+  def zip[B](xs: Iterable[B]): CC[(A @uncheckedVariance, B)] = zipWith(xs)((_, _))
+
   // sound bcs of VarianceNote
```
This comment is really related to the `@uncheckedVariance` above, you should not insert a new line in between them.

marcelocenerine replied:
ooops....will do

julienrf commented on src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -835,9 +835,24 @@ trait IterableOps[+A, +CC[X], +C] extends Any {
+    *  @param   f   The function to apply to each pair of elements
+    *  @tparam  B     the type of the elements in the second half of the combined pairs
+    *  @tparam  R     the type of the elements in the resulting collection
+    *  @return        a new collection of type `That` containing the results of applying the given function `f`
```
We should not use `That`, that’s a left over of the standard collections based on `CanBuildFrom` :)

We should just say “a new $Coll containing …”

julienrf replied:
Could you also fix `zip`’s scaladoc at the same time?

marcelocenerine replied:
that's good to know. I was in doubt about it. I guess you meant `$Coll` with lowercase `c`, right?

julienrf replied:
Yeah, `$coll` with lowercase `c` is better, you’re right.

julienrf commented on src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -835,9 +835,24 @@ trait IterableOps[+A, +CC[X], +C] extends Any {
+    *                 to each pair of corresponding elements of this $coll and `that`. The length
+    *                 of the returned collection is the minimum of the lengths of this $coll and `that`.
+    */
+  def zipWith[B, R](xs: Iterable[B])(f: (A, B) => R): CC[R] = fromIterable(View.ZipWith(toIterable, xs, f))
```
Could you rename the `xs` parameter to `that`, to match the scaladoc?

marcelocenerine replied:
will do

julienrf commented on src/main/scala/strawman/collection/View.scala:
```diff
@@ -183,11 +183,11 @@ object View extends IterableFactory[View] {
-  case class Zip[A, B](underlying: Iterable[A], other: Iterable[B]) extends View[(A, B)] {
-    def iterator() = underlying.iterator().zip(other)
+  /** A view that generalizes the zip operation by applying a function to each pair of elements
+    * in the underlying collection and another collection or iterator.
```
We should remove “or iterator” since this is not anymore the case

marcelocenerine replied:
will do

julienrf commented on test/junit/src/test/scala/strawman/collection/ArrayOpsTest.scala:
```diff
@@ -0,0 +1,71 @@
+    val array: ArrayOps[Int] = Array(1, 2, 3)
+    val result: Array[(Int, String)] = array.zip(List("a", "b", "c"))
+
+    assertArrayEquals(Array((1, "a"), (2, "b"), (3, "c")).asInstanceOf[Array[AnyRef]], result.asInstanceOf[Array[AnyRef]])
```
Isn’t it working if you omit the `asInstanceOf`s?

marcelocenerine replied:
no, it does not compile:
```
overloaded method value assertArrayEquals with alternatives:
[error]   (x$1: Array[Long],x$2: Array[Long])Unit <and>
[error]   (x$1: Array[Int],x$2: Array[Int])Unit <and>
[error]   (x$1: Array[Short],x$2: Array[Short])Unit <and>
[error]   (x$1: Array[Char],x$2: Array[Char])Unit <and>
[error]   (x$1: Array[Byte],x$2: Array[Byte])Unit <and>
[error]   (x$1: Array[Object],x$2: Array[Object])Unit
[error]  cannot be applied to (Array[(Int, String)], Array[(Int, String)])
[error]     assertArrayEquals(Array((1, "a"), (2, "b"), (3, "c")), result)
[error]     ^
```

julienrf replied:
What if you replace that with `assertTrue(Array((1, "a"), ...).equals(result))`?

marcelocenerine replied:
`equals` is not overwritten so it does not compare the content of `Array` objects. But I can do what you suggested with `sameElements` so that these asserts become shorter.

julienrf commented on test/junit/src/test/scala/strawman/collection/immutable/TreeSetTest.scala:
```diff
@@ -17,4 +17,60 @@ class TreeSetTest {
     assertEquals(set, set drop Int.MinValue)
```
I don’t think we need these tests since `TreeSet` inherits the default implementation, which is already tested in `IterableViewLikeTest`.

marcelocenerine replied:
`TreeSet` has overloaded `zip` and `zipWith` methods inherited from `SortedSet`. These tests are testing the overloaded methods.

julienrf replied:
Oh yes, that’s right.

julienrf reviewed (changes requested):
Thanks a lot @marcelocenerine! Overall that looks great, I’ve left a few comments, mainly about documentation or typos.

Could you also fix the merge conflicts?

odersky commented:
My proposal would be to find a way to have any additional methods sit somewhere where we can easily decide where to put them in the end. There is a case for each individual method, but there is also a clear case _against_ adding dozens of new methods. The trick is to balance one against the other. In the case of `zipWith` I am particularly sceptical because

 - we already had it in early versions of Scala and removed it
 - it can be expressed as `(xs, ys).zipped.map`, which was the primary reason to remove it.

julienrf commented:
> it can be expressed as `(xs, ys).zipped.map`, which was the primary reason we removed it.

Wow, today I learn about `(xs, ys).zipped`! Indeed, we can achieve exactly the same result as `zipWith`, and we can even do more because `Tuple2Zipped` also has `filter`, `flatMap` and a few others.

Maybe it’s just me but I really think the discoverability of this `(xs, ys).zipped` thing is very low: if you have two collections the only chance to know about this method is to go to the `Tuple2` scaladoc page: http://scala-lang.org/api/current/scala/Tuple2.html. But if you have three collections there is no way to know about `Tuple3Zipped.Ops#zipped` unless you know this class name because there is no `Tuple3` entry in the stdlib scaladoc.

One way to make things more discoverable would be to make `zipped` an actual operation (not a decorator) of the collections. We could maybe name it `zipView` to emphasize the fact that it doesn’t return a collection: `xs.zipView(ys)`.

julienrf commented:
@marcelocenerine @Ichoran @szeiger Do you have an opinion about the `viewWith` operation described above? I discussed with @odersky and he told me that he thinks we should keep one way of doing things, `(xs, ys).zipped`, and simply improve its discoverability by defining a link to this method in `zip` and `zipWithIndex` scaladoc. I do think that it would be better to move to `viewWith` instead and deprecate `zipped`, because I think it would be even better to have a `viewWith` entry in the scaladoc of the collections and because decorators (which is the case of `zipped`) don’t play well with type inference.

odersky commented:
The problem with `zipWith, viewWith` etc is that 

 - it is only binary
 - it can only fo map, not flatMap, filter, or foreach.

`zipped` is a lot of more flexible. I don't see a reason why we should give that up!

julienrf commented:
No, sorry I wasn’t clear:

~~~ scala
(xs viewWith ys).map(binaryFunction) == (xs, ys).zipped.map(binaryFunction)
(xs viewWith ys).forall(binaryPredicate) == (xs, ys).zipped.forall(binaryPredicate)
(xs viewWith ys viewWith zs).filter(ternaryPredicate) == (xs, ys, zs).zipped.filter(ternaryPredicate)
~~~

odersky commented:
@julienrf Ah, I misunderstood indeed. But then there's still the problem that it's binary only. I have used `zipped` also for 3 and 4 tuples.

Ichoran commented:
I would prefer staying with `zipped` because (1) it's already there, and (2) we can define it for as many arities as we want.  If it produces a proper view, though, that would be a nice improvement.

julienrf commented:
> (1) it's already there

My point is that it’s hard to know that it exists and that a `viewWith` (or whatever its name) method would be easier to work with.

> (2) we can define it for as many arities as we want

`viewWith` can do that too:

~~~ scala
(xs1 wiewWith xs2 viewWith xs3 viewWith xs4).filter((x1, x2, x3, x4) => x1 == x2 && x3 != x4)
~~~

odersky commented:
> viewWith can do that too:

    (xs1 viewWith xs2 viewWith xs3 viewWith xs4).filter((x1, x2, x3, x4) => x1 == x2 && x3 != x4)

I can see how that works if tuples are HLists, but otherwise don't you need to write:

    (xs1 viewWith xs2 viewWith xs3 viewWith xs4).filter(((x1, x2), x3), x4) => x1 == x2 && x3 != x4)

(and incur the expense of constructing and deconstructing the intermediare tuples)?

julienrf commented:
~~~ scala
trait Iterable[A] {
  def viewWith[B](that: Iterable[B]): View2[A, B] = new View2[A, B](this, that)
}

class View2[A, B](as: Iterable[A], bs: Iterable[B]) {
  def viewWith[C](cs: Iterable[C]): View3[A, B, C] = new View3(as, bs, cs)
  def map[C](f: (A, B) => C): Iterable[C] = …
  // etc. (other methods omitted)
}

class View3[A, B, C](as: Iterable[A], bs: Iterable[B], cs: Iterable[C]) {
  def viewWith[D](ds: Iterable[D]): View4 = … // you got the idea, right? We could support as many arities as we want
  def map[D](f: (A, B, C) => D): Iterable[D] = …
}
~~~

Scalaz used to have a similar operator named `|@|`. Play-json uses `and`. But they all do the same thing in essence.

odersky commented:
@julienrf You are right, that solves the problem. I agree that's more regular than `zipped`, which is kind of special because it is defined on tuple classes and therefore is hard to discover. So I think I have run out of objections against deprecating `zipped` in favor of `viewWith` 😄 .

marcelocenerine commented:
@julienrf, I like your suggestion and it seems to solve both problems. 

How many arities should be supported? `Tuple*Zipped` goes up to 3

Wouldn't `zipView` sound more intuitive? At least the result for `xs zipView ys` would be similar to what we get with `xs zip yz`: `(x1, y1)`. On the other hand, for other arities there would be a mismatch: `zipView`: `(x1, y1, z1)` vs. `zip`: `((x1, y1), z1)`.

So are we going to abandon this PR and have a new issue filed for `viewWith` (or some other name)? Or should I just update this PR?

Ichoran commented:
@julienrf - Creating an N-ary zip requires N object creations with N*(N-1) assignments, instead of 1 and N in this scheme.  If it can be inlined away, that's fine.  But it should be annotated @inline, and we should check whether inlining actually results in efficient code.  Alternatively, if we leave it only going up to 3, it's okay.

Also, when  you run out of the predefined `ViewN`s, it's going to go back to the `Iterable` one, which will be a bit weird, but may be okay.  It should be documented, though, because it's not just a compiler error any more.  We're more likely to hit this if we leave the limit at 3.

odersky commented:
The `viewWith` scheme looks a it simpler than the previous `zipped` scheme,
in parrticular since there's no need for custom implementations of all
supported methods. So I think we can probably go up to arity 4. That should
be enough for anyone (famous last words).

 - Martin

On Fri, Sep 22, 2017 at 3:35 AM, Ichoran <notifications@github.com> wrote:

> @julienrf <https://github.com/julienrf> - Creating an N-ary zip requires
> N object creations with N*(N-1) assignments, instead of 1 and N in this
> scheme. If it can be inlined away, that's fine. But it should be annotated
> @inline <https://github.com/inline>, and we should check whether inlining
> actually results in efficient code. Alternatively, if we leave it only
> going up to 3, it's okay.
>
> Also, when you run out of the predefined ViewNs, it's going to go back to
> the Iterable one, which will be a bit weird, but may be okay. It should
> be documented, though, because it's not just a compiler error any more.
> We're more likely to hit this if we leave the limit at 3.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/collection-strawman/pull/223#issuecomment-331326374>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAwlVrdlGqIuo98MkoqvitzpsTm81t3dks5skw7fgaJpZM4PK6dW>
> .
>



-- 

Martin Odersky
EPFL and Lightbend

julienrf commented:
@Ichoran 

> Also, when you run out of the predefined ViewNs, it's going to go back to the Iterable one, which will be a bit weird, but may be okay. It should be documented, though, because it's not just a compiler error any more. We're more likely to hit this if we leave the limit at 3.

What do you mean by “it’s going back to the `Iterable` one”? If `View4` has no `viewWith` method, then trying to view 5 collections will result in a compile-time error.

@odersky 

> The `viewWith` scheme looks a it simpler than the previous `zipped` scheme, in parrticular since there's no need for custom implementations of all supported methods.

I’m not sure to follow you: what I named `View2` in my previous comment was equivalent to the current `Tuppled2Zipped`, `View3` is equivalent to the current `Tuppled3Zipped`. These classes have a custom implementation of `map`, `filter`, etc. If we really want to not have special classes for that we need to use an `HList` to actually abstract over the arity of the view, but that would go against our purpose since we want to avoid unnecessary tupling and untupling when `map`ping (or, in general, transforming) several collections element-wise.

@marcelocenerine 

`viewWith` reads slightly better for me. You can update this PR or create a new one, as you prefer, since the new solution will be quite different :)

odersky commented:
@julienrf I was hoping that the `ViewN` classes, being regular views, would not need custom implementations of `map`, `flatMap`, `filter`, `foreach`. But I now see that's impossible because the arity of these functions is wrong. Still, maybe go up to 4?

Ichoran commented:
@julienrf - I was assuming there'd be an implicit conversion from View2[A, B] to Iterable[(A, B)] and so on with higher arities so that it would act more like a regular View.  Not sure why I assumed that.  If not, it gets around that problem.  I'm not _entirely_ sure that `View2` is a good name, though.  `map` and friends are eager (not lazy like on `View`), and only a small subset of `Iterable` is available.

julienrf commented:
> I'm not entirely sure that View2 is a good name, though. map and friends are eager (not lazy like on View), and only a small subset of Iterable is available.

Yeah, that’s a good point, do you have another name suggestion? Do you think it would make sense to make operations return a `View` instead of a strict collection?

odersky commented:
I agree, they should not be called `View` because they don't represent normal views. They only delay the zipping, not the next operation after it. In light of this, I also think `withView` as a method name is misleading. How about `zipped` but now written infix? I.e.

    xs.zipped(ys).zipped(zs).map((x, y, z) => ...)

julienrf commented:
I’m not a big fan of `zipped`, I think the name doesn’t say anything about the difference with `zip`.

marcelocenerine commented:
`lazyZip`?

I have a few questions:
- what operations beside `map`, `flatMap` and `filter` should be supported on the `ViewN` (actual name TBD) classes? The existing `TupleNZipped` classes also define `exists`, `forall` and `foreach`. There is also an implicit conversion to `Traversable` (`zippedTraversableNToTraversable`) in each of them which adds a plenty of other methods.
- what should be the return type of `filter`? another `ViewN`, `Iterable[(A, B, ..., N)]`, ... The existing`TupleNZipped` classes return `TupleN[ColA, ColB, ..., ColN]`, which unzips the collections:

```
scala> (Set(1, 2, 3, 4, 5, 6), List("a", "b", "c", "d", "e", "f", "g")).zipped
   .filter((a, b) => a % 2 == 0)
res6: (scala.collection.immutable.Set[Int], List[String]) = (Set(6, 2, 4),List(c, d, f))
```

julienrf commented:
If we make it an actual view (which is not the _completely_ case of the current `Tuple2Zipped`), then indeed return types should be views everywhere and people would have to force it (with `.to`) to evaluate it.

The current behaviour of `Tuple2Zipped` is to have strict `map`, `flatMap` and `filter` operations but the conversion to `Traversable` doesn’t evaluate its elements immediately (ie, it’s lazy).

I think it would be more consistent to make it an actual view (`map`, `flatMap` and `filter` would return a `View2[A, B]`, and have an implicit conversion to a `View[(A, B)]`). But then people would have to write the following:

~~~ scala
(xs viewWith ys).map(f).to(xs.iterableFactory)
~~~

Instead of just:

~~~ scala
xs.zipWith(ys)(f)
~~~

As we wanted to do, initially, in this PR.

odersky commented:
OK, let's stick with zipWith (up to arity 3 or 4) then.

On Mon, Oct 2, 2017 at 9:46 AM, Julien Richard-Foy <notifications@github.com
> wrote:

> If we make it an actual view (which is not the *completely* case of the
> current Tuple2Zipped), then indeed return types should be views
> everywhere and people would have to force it (with .to) to evaluate it.
>
> The current behaviour of Tuple2Zipped is to have strict map, flatMap and
> filter operations but the conversion to Traversable doesn’t evaluate its
> elements immediately (ie, it’s lazy).
>
> I think it would be more consistent to make it an actual view (map,
> flatMap and filter would return a View2[A, B], and have an implicit
> conversion to a View[(A, B)]). But then people would have to write the
> following:
>
> (xs viewWith ys).map(f).to(xs.iterableFactory)
>
> Instead of just:
>
> xs.zipWith(ys)(f)
>
> As we wanted to do, initially, in this PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/collection-strawman/pull/223#issuecomment-333462741>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAwlVjKhn9zU_7_PRzawAQMWKF-1Rt79ks5soJTOgaJpZM4PK6dW>
> .
>



-- 

Martin Odersky
EPFL and Lightbend

marcelocenerine commented:
@julienrf, does it mean we are back to the original plan?

julienrf commented:
@marcelocenerine I still think that the current `(xs, ys).zipped` thing is not the best way approach because of its poor discoverability. Maybe `lazyZip` is a better name than `viewWith` since it doesn’t contain a (misleading) “view”.

So, let’s add a `lazyZip` method that returns a `Tuple2Zipped` instance, which in turn contains a `lazyZip` method returning a `Tuple3Zipped` instance. We can go up to `Tuple3Zipped`. (You can not reuse the current `TupleNZipped` implementations since they are in scala-library, you’ll have to re-implement them in the strawman).

What do you think?

marcelocenerine commented:
@julienrf, thanks for clarifying. It sounds good to me. I will update the PR in the coming days.

odersky commented:
Ah I misread that. I guess I wanted

    xs.zipWith(ys).zipWith(zs).map(f)

But that's also not ideal because `zipWith` has different signatures elsewhere.

I don't think after all zipWith should give a full view. Ideally, only `view` should produce a view. Maybe, `zipWith` called on a view `xs` could yield again a view. Not sure how hard that is to achieve.

So I guess our options are:

 - a viewWith method that produces a full view (if we can achieve that without too much effort)

or/and

 - The previous (a, b, c).zipped scheme
 - An infix operator instead of zipped. In principle, infix is nicer, but the difficulty is to come up with a name. I still think `zipped` is also OK as an infix operator name.

marcelocenerine commented:
@julienrf, could you take a quick look to see if this is what you had in mind? (this is the 3rd option @odersky suggested above - except the name is still TBD): https://github.com/marcelocenerine/collection-strawman/commit/152db8b65203d7aa5b20cb1c3fe12bab092c08d1

Questions:
- currently `map`/`flatMap`/`filter` have `Iterable[_]` as return type (although returned collection is == `as`'s type). Is there a way to fix this (maybe with `BuildFrom`?) so that the new `TupleNZipped` types match the existing ones in the Scala library?
- should we have implicit conversions from `TupleNZipped` to `Iterable[TupleN]`?

julienrf commented:
Hey @marcelocenerine, thanks for this work!

> currently map/flatMap/filter have Iterable[_] as return type (although returned collection is == as's type). Is there a way to fix this (maybe with BuildFrom?) so that the new TupleNZipped types match the existing ones in the Scala library?

Yes, you should use [`BuildFrom`](https://static.javadoc.io/ch.epfl.scala/collection-strawman_2.13.0-M2/0.5.0/strawman/collection/BuildFrom.html). The current signature of `Tuple2Zipped` is:

~~~ scala
class Tuple2Zipped[El1, Repr1, El2, Repr2]
~~~

I would name the new one as follows:

~~~ scala
class Tuple2Zipped[El1, C1, El2, C2]
~~~

Then, the migration scheme for `map`, `flatMap` would be the following:

~~~ scala
// current
def map[B, To](f: (El1, El2) ⇒ B)(implicit cbf: CanBuildFrom[C1, B, C]): C 
// strawman
def map[B, C](f: (El1, El2) ⇒ B)(implicit bf: BuildFrom[C1, B, C]): C
~~~

Also, you should use `bf.fromSpecificIterable` rather than `bf.newBuilder()` to build your result:

~~~ scala
def map[B, C](f: (El1, El2) ⇒ B)(implicit bf: BuildFrom[C1, B, C]): C =
  bf.fromSpecificIterable(coll1)(new View[B] {
    def iterator() = …
  })
~~~

> should we have implicit conversions from TupleNZipped to Iterable[TupleN]?

Yes, I would keep that feature too.

marcelocenerine commented:
@julienrf, awesome!
Just out of curiosity: what is the reason for using `bf.fromSpecificIterable` in place of `bf.newBuilder()`? Is it just to make `map`/`flatMap` lazy?

I updated my branch, but using `bf.newBuilder()`...please see what you think: https://github.com/marcelocenerine/collection-strawman/commit/d1ca265dd1d79b0c0e4161c9a97c73896ee89cc0

Thank you

julienrf commented:
> Just out of curiosity: what is the reason for using bf.fromSpecificIterable in place of bf.newBuilder()?

It’s to preserve laziness of the source collection. If you write `(xs lazyZip ys).map(f)` and that `xs` is a `View` you want the result to be a `View` that hasn’t evaluated its elements yet. You can find more details about that in the [DESIGN.md](https://github.com/scala/collection-strawman/blob/master/documentation/DESIGN.md#view-based-default-implementations) file.

The code you pushed looks good excepted for the use of `newBuilder` and also I think that instead of taking implicit conversions from `C => Iterable[A]` you should just put an upper bound to the type parameter `C`:

~~~ scala
class Tuple2Zipped[A1, C1 <: Iterable[A1], A2, C2 <: Iterable[A2]](c1: C1, c2: C2) { … }
~~~

Does that make sense?

marcelocenerine commented:
>It’s to preserve laziness of the source collection

I presumed that was the reason. Thanks for pointing me to the design page which answered some other questions I had.
The `map`/`flatMap`/`filter` methods are now using `bf.fromSpecificIterable`. `filter` is trickier as the result is a tuple containing each individual collection whose elements are filtered (not sure how useful this is). I'm not pretty sure my solution for preserving laziness is the most efficient one.  I can see at least one issue with GC given that each individual view in the resulting tuple holds references to all collections in the source `TupleNZipped`

>instead of taking implicit conversions from `C => Iterable[A]` you should just put an upper bound to the type parameter `C`

In fact this was my initial implementation but it took long until I got it right (especially the `Iterable.lazyZip` method). I'm still using an implicit conversion in `Iterable.lazyZip` to convert `C` to `Iterable[A]`.

Here is the latest code: https://github.com/scala/collection-strawman/compare/master...marcelocenerine:lazyZip

One thing that I noticed that seemed a bit odd is that the implicit conversion from `Tuple4Zipped` to `Iterable[(A, B, C, D)]` allows the `lazyZip` method to be called on the resulting iterable, which gives back something like `Tuple2Zipped[(A, B, C, D), List[(A, B, C, D)], E, List[E]]`. If this is not desirable, then I think we may need to remove the implicit conversion from `Tuple4Zipped`.

If you don't see any other major issues, I can work on the scaladocs and update this PR shortly.

Thanks for the help so far, @julienrf!

julienrf commented:
> filter is trickier as the result is a tuple containing each individual collection whose elements are filtered. I'm not pretty sure my solution for preserving laziness is the most efficient one in that case.

Hmm, indeed… This is one requires two traversals, like `partition` does (see [here](https://github.com/scala/collection-strawman/blob/master/collections/src/main/scala/strawman/collection/View.scala#L135-L152)). By the way, I find it weird that, in general, `(xs lazyZip ys).xxx(f)` is equivalent to `(xs zip ys).xxx(f.tupled)` excepted when `xxx` is `filter`, where it is equivalent to `(xs zip ys).filter(f.tupled).unzip` (note the final `unzip` call!). What do you think if we change the behaviour of this method to return a single collection of tuples, instead?

~~~ scala
def filter[C](p: (El1, El2) => Boolean)(implicit bf: BuildFrom[C1, (El1, El2), C]): C
~~~

> One thing I noticed that seemed a bit odd is that the implicit conversion from Tuple4Zipped to Iterable[(A, B, C, D)] allows the lazyZip method to be called on the resulting iterable

Oh, this is not desirable, indeed :(

I do think that the conversion to `Iterable` is still usefull, though. Another workaround would be to make the `lazyZip` operation available via an implicit conversion from `Iterable`. Then `Tuple4Zipped` wouldn’t trigger that implicit conversion since it’s not an `Iterable`:

~~~ scala
object Iterable {
  implicit class LazyZipOps[A, CA <: Iterable[A]](`this`: CA) {
    def lazyZip[B, CB <: Iterable[B]](that: CB): Tuple2Zipped[A, CA, B, CB] = …
  }
}
~~~

BTW, I realize that what started as a minor issue (adding a `zipWith` operation) turned to be a huge thing. Just let me know if you don’t have lot of time, maybe we can find a way to share the work. In any case, thanks a lot for your help!

marcelocenerine commented:
>I find it weird that, in general, (xs lazyZip ys).xxx(f) is equivalent to (xs zip ys).xxx(f.tupled) excepted when xxx is filter, where it is equivalent to (xs zip ys).filter(f.tupled).unzip (note the final unzip call!)

Exactly! I mentioned that in one of my previous comments. I changed the implementation as you suggested, which makes more sense to me.

>Another workaround would be to make the lazyZip operation available via an implicit conversion from Iterable

I like that. It also simplifies the signature of the first `lazyZip` method and avoids the `@uncheckedVariance` annotations. I slightly changed the signature you proposed to preserve the type information of the collection elements.

>I realize that what started as a minor issue (adding a zipWith operation) turned to be a huge thing. Just let me know if you don’t have lot of time, maybe we can find a way to share the work. In any case, thanks a lot for your help!

It's my pleasure to contribute :). I think now it should pretty much done (or at least close). If you see any room for improvement or want to make any tweaks, feel free to cherry-pick my commit and work on top of it. In fact next week my free time will be limited, so I may not be able to address review comments within a day or two.

The PR is now updated. I have attached the [benchmarks results](https://github.com/scala/collection-strawman/files/1385316/benchmark_zip_lazyZip.pdf) of `zip` and `lazyZip`. In general `(xs lazyZip ys).map((_, _))` slightly underperforms `(xs zip ys)`, except in `ImmutableArray`, `TreeSet` and `ArrayBuffer` where the difference was more significant (especially in `ImmutableArray` where `map` is overridden for efficiency). This is also a trend in the Scala std library (`zip` vs `zipped`) though. I ran the benchmarks once again and the results for `TreeSet` were better than in the first time.

marcelocenerine commented on benchmarks/time/src/main/scala/strawman/collection/immutable/HashSetBenchmark.scala:
```diff
@@ -136,6 +136,15 @@ class HashSetBenchmark {
+  def transform_lazyZip(bh: Blackhole): Unit = bh.consume(xs.lazyZip(xs).map((_, _)))
+
+  @Benchmark
   def transform_unzip(bh: Blackhole): Unit = bh.consume(zipped.unzip)
```
is there any recommendation for when (not) to add new benchmarks to the project? These are going to add several hours to the running time

julienrf replied:
No recommendation so far. New benchmarks are always welcomed for now. In my case anyway I only run a subset of the benchmarks when I run them.

julienrf replied:
Also, what I do sometimes is that I group related benchmarks together so that it’s simpler to run just the benchmarks that I’m interested in: https://github.com/scala/collection-strawman/blob/master/benchmarks/time/src/main/scala/strawman/collection/GroupMapBenchmarks.scala

julienrf commented on benchmarks/time/src/main/scala/strawman/collection/immutable/HashSetBenchmark.scala:
```diff
@@ -136,6 +136,15 @@ class HashSetBenchmark {
+  def transform_zipWithIndex(bh: Blackhole): Unit = bh.consume(xs.zipWithIndex)
+
+  @Benchmark
+  def transform_lazyZip(bh: Blackhole): Unit = bh.consume(xs.lazyZip(xs).map((_, _)))
```
Could you also add more benchmarks to compare `xs.lazyZip(ys).map(f)` with `xs.zip(ys).map(f.tupled)`?

julienrf commented on collections/src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -878,7 +877,30 @@ trait IterableOps[+A, +CC[X], +C] extends Any {
+      *    {{{
+      *    val xs = $Coll(1, 2, 3)
+      *    val res = (xs lazyZip xs lazyZip xs lazyZip xs).map((a, b, c, d) => a + b + c + d)
+      *    // res == $Col(4, 8, 12)
```
Typo: $Col**l**

julienrf replied:
Or maybe you should just use `List` here.

julienrf commented on collections/src/main/scala/strawman/collection/Iterable.scala:
```diff
@@ -878,7 +877,30 @@ trait IterableOps[+A, +CC[X], +C] extends Any {
+      * @return a decorator `Tuple2Zipped` that allows strict operations to be performed on the lazily evaluated pairs
+      *         or chained calls to `lazyZip`. Implicit conversion to `Iterable[(A, B)]` is also supported.
+      */
+    def lazyZip[B, C2[X] <: Iterable[X]](that: C2[B]): Tuple2Zipped[A, C1[A], B, C2[B]] = new Tuple2Zipped(`this`, that)
```
By convention we use the `CC` identifier for **C**ollection type **C**onstructors, while `C` is used for **C**ollection types. BTW, I’m wondering, here, what’s the advantage of using a collection type constructor over a collection type?

marcelocenerine replied:
>I’m wondering, here, what’s the advantage of using a collection type constructor over a collection type?

```
scala> import strawman.collection._
import strawman.collection._

scala> import strawman.collection.immutable._
import strawman.collection.immutable._

scala> val xs: List[Int] = List(1, 2, 3)
xs: strawman.collection.immutable.List[Int] = List(1, 2, 3)

scala> xs.lazyZip(xs)
<console>:19: error: value lazyZip is not a member of strawman.collection.immutable.List[Int]
       xs.lazyZip(xs)
          ^
```

Using collection type in `TupleNZipped.lazyZip`:

```
scala> xs.lazyZip(xs).lazyZip(xs)
<console>:19: error: inferred type arguments [Nothing,strawman.collection.immutable.List[Int]] do not conform to method lazyZip's type parameter bounds [B,C3 <: strawman.collection.Iterable[B]]
       xs.lazyZip(xs).lazyZip(xs)
                      ^
<console>:19: error: type mismatch;
 found   : strawman.collection.immutable.List[Int]
 required: C3
       xs.lazyZip(xs).lazyZip(xs)
                              ^
```

I got confused with this one as well. I started with the signature suggested in one of your last comments, but then I faced these errors. This was pretty much the reason I used implicit conversions in the first iteration. Am I missing anything?

julienrf replied:
Hmm ok I see. Thanks for the explanation!

julienrf commented on collections/src/main/scala/strawman/collection/TupleZipped.scala:
```diff
@@ -0,0 +1,347 @@
+    def next() = (elems1.next(), elems2.next())
+  }
+
+  def className = getClass.getName
```
I’m not sure the class name `Tuple2Zipped` is very useful. What do you think of implementing `toString` as follows?

~~~ scala
def toString = s"$coll1 lazyZip $coll2"
~~~

marcelocenerine replied:
ok, will do

julienrf commented on collections/src/main/scala/strawman/collection/TupleZipped.scala:
```diff
@@ -0,0 +1,347 @@
+object Tuple2Zipped {
+  implicit def tuple2ZippedToIterable[El1, C1 <: Iterable[El1],
+                                      El2, C2 <: Iterable[El2]](zipped2: Tuple2Zipped[El1, C1, El2, C2]): Iterable[(El1, El2)] =
+    new View[(El1, El2)] { def iterator() = zipped2.iterator() }
```
Maybe we should be more explicit and also put `View[(El1, El2)]` as the static return type?

marcelocenerine replied:
ok, will do

julienrf commented on test/junit/src/test/scala/strawman/collection/TupleZippedTest.scala:
```diff
@@ -0,0 +1,179 @@
+  def tuple4Zipped_empty(): Unit = {
+    assertTrue(zipped3.lazyZip(List.empty).isEmpty)
+  }
+}
```
Can you also check that `TreeSet(1, 2, 3).lazyZip(List(4, 5, 6)).map((x, y) => x + y)` returns a `TreeSet`?

And that `TreeMap(1 -> "foo", 2 -> "bar").lazyZip(List("baz", "bah")).map { case ((k, v), s) => k -> (v ++ s) }` also returns a `TreeMap`?

marcelocenerine replied:
it does not work well with `TreeMap` :(

marcelocenerine replied:
The issue here doesn't only concern `TreeMap` but `Map`s in general. I guess the reason is the same for having overloaded methods in `Map` (e.g. `map`, `flatMap`, `concat`).

Having a more specific implicit conversion in the `Map` companion object may address this:

```
implicit class LazyZipOps[K, V, CC1[K, V] <: Map[K, V]](`this`: CC1[K, V]) {
  def lazyZip[B, CC2[B] <: Iterable[B]](that: CC2[B]): Tuple2Zipped[(K, V), CC1[K, V], B, CC2[B]] = new Tuple2Zipped(`this`, that)
}
```

julienrf replied:
Yes, that’s a good idea.

Also, I think we don’t need to keep track of the precise type of the zipped collections: we just need `C1` (to drive the type of the collection resulting from a transformation operation applied to the zipped collections), but we never need `C2`, `C3` or `C4`.

So, I think we could simplify things as follows:

~~~ scala
// TupleZipped.scala
final class Tuple2Zipped[El1, El2, C1 <: Iterable[El1]](coll1: C1, coll2: Iterable[El2]) { … }
final class Tuple3Zipped[El1, El2, El3, C1 <: Iterable[El1]](coll1: C1, coll2: Iterable[El2], coll3: Iterable[El3]) { … }
// same for Tuple4Zipped

class LazyZipOps[A, C1 <: Iterable[A]](`this`: C1) {
  def lazyZip[B](that: Iterable[B]): Tuple2Zipped[A, B, C1] = new Tuple2Zipped[A, B, C1](`this`, that)
}

// Iterable.scala
object Iterable {
  implicit def toLazyZipOps[CC[X] <: Iterable[X], A](it: CC[A]): LazyZipOps[A, CC[A]] = new LazyZipOps(it)
}

// Map.scala
object Map {
  implicit def toLazyZipOps[CC[X, Y] <: Iterable[(X, Y)], K, V](it: CC[K, V]): LazyZipOps[(K, V), CC[K, V]] = new LazyZipOps(it)
}
~~~

marcelocenerine replied:
yeah, I was considering doing that too. The collection types were needed in previous iterations when `filter` was equivalent to `(xs zip ys).filter(f.tupled).unzip`, but now I can get rid of them

julienrf reviewed (commented):
That looks great, thanks! I’ve left some comments.

julienrf commented:
Hey @marcelocenerine what’s your status on this work? This really looks great but needs a few tweaks. If you are busy with other things I can try to finish it. Just let me know what works best for you :)

marcelocenerine commented:
Hi @julienrf , sorry for the delay. I was away last week and didn't find time to address all the comments yet.
I have almost everything sorted. The only things left are:
- benchmarks to compare `xs.lazyZip(ys).map(f)` with `xs.zip(ys).map(f.tupled)`
- make sure `TreeMap(1 -> "foo", 2 -> "bar").lazyZip(List("baz", "bah")).map { case ((k, v), s) => k -> (v ++ s) }` also returns a TreeMap

I left a comment regarding the latter. Please see what you think. I believe I can finish it and update the PR this evening. But if you need it earlier than that, please feel free to cherry pick my commit and work on top of that. I can push the changes I have made so far...

julienrf commented:
OK, thanks for your detailed response! So, I will let you finish :)

marcelocenerine commented:
@julienrf, all the review comments are now addressed. Many thanks for your suggestions.

I ended up renaming the `TupleNZipped` classes to `LazyZipN`. The naming in the Scala std library is likely influenced by the fact that `TupleNZipped` objects are created out of tuples, which is no longer the case in strawman. I think that would be misleading given that all the operations available on the new classes don't create any tuples at all (except the implicit conversion to `Iterable[(A, B)]`). Maybe the scaladocs need some tweaks to make this clear. I'm not totally convinced that `LazyZipN` is a good name, so please let me know if you have a better suggestion or if you don't agree in renaming them.

I'm running the benchmarks again and will post the results as soon as they complete.

julienrf commented on collections/src/main/scala/strawman/collection/LazyZipOps.scala:
```diff
@@ -0,0 +1,352 @@
+import scala.{Boolean, StringContext, Unit}
+import scala.language.implicitConversions
+
+final class LazyZipOps[A, C1 <: Iterable[A]] private[collection](`this`: C1) {
```
We could even make it a value class (even though I bet the saving wouldn’t be significant at all)

marcelocenerine replied:
this is a very easy change, so I think it's worth doing it despite how much saving we eventually get.

julienrf reviewed (approved):
Awesome contribution @marcelocenerine, thanks a lot!

marcelocenerine commented:
Please find attached the [benchmark results](https://github.com/scala/collection-strawman/files/1421494/benchmarks_lazyZip_20171026.pdf) comparing `xs.lazyZip(ys).map(f)` with `xs.zip(ys).map(f.tupled)`

julienrf commented:
@marcelocenerine That looks great to me, could you please squash your commits into a single one so that I can merge the PR?

marcelocenerine committed:
Add lazyZip operation

marcelocenerine commented:
>for `ImmutableArray`, if you have less than 1000 elements, `xs.zip(ys).map(f)` is faster than `xs.lazyZip(ys).map(f)`

I figured there was a problem there. The views returned by `LazyZipN.map` weren't overriding `knownSize`. Now `xs.lazyZip(ys).map(f)` is faster than `xs.zip(ys).map(f)` for `ImmutableArray`. Thanks for spotting that.

![image](https://user-images.githubusercontent.com/1107367/32180851-fbc7f70e-bd8a-11e7-9ff1-45df8ac62fbb.png)
![image](https://user-images.githubusercontent.com/1107367/32180871-07c3bbd8-bd8b-11e7-9a25-c6c0dc40b34c.png)


My commits are now squashed.

julienrf commented on collections/src/main/scala/strawman/collection/LazyZipOps.scala:
```diff
@@ -0,0 +1,356 @@
+
+object LazyZip2 {
+  implicit def lazyZip2ToIterable[El1, El2](zipped2: LazyZip2[El1, El2, _]): View[(El1, El2)] =
+    new View[(El1, El2)] { def iterator() = zipped2.iterator() }
```
You should also override `knownSize` here, then:

~~~ scala
override def knownSize = zipped2.coll1.knownSize min zipped2.coll2.knownSize
~~~

And same for other arities.

marcelocenerine replied:
oops. done

julienrf commented:
Woohoo, thanks a lot Marcelo!
